### ISSUE TITLE
String Parameter and Int Parameter combination causes InvalidProtocolBufferException

### ISSUE THREAD

**Issue by dwdii** [Azure/hdinsight-phoenix-sharp#8]
I have been successful with one or more parameters that are data type `string` only, but when I introduce an integer as a parameter, I receive a ProtoBuf exception. My unit test's stack trace is shown below.  I've looked through the unit tests for this project and some usage from the twitter streaming project, but have only found examples of `string` only parameter sets. 

My unit test code (CommandPrepareWith1String1IntParamsTest) can be found in my repo at: https://github.com/dwdii/GarudaUtil/blob/master/Garuda.Data.Test/PhoenixUnitTest.cs

The error message is:

Google.Protobuf.InvalidProtocolBufferException: While parsing a protocol message, the input ended unexpectedly in the middle of a field.  This could mean either than the input has been truncated or that an embedded message misreported its own length.

```
Result StackTrace:	
at Google.Protobuf.CodedInputStream.RefillBuffer(Boolean mustSucceed)
   at Google.Protobuf.CodedInputStream.ReadRawByte()
   at Google.Protobuf.CodedInputStream.SlowReadRawVarint32()
   at Google.Protobuf.CodedInputStream.ReadRawVarint32()
   at Google.Protobuf.CodedInputStream.ReadString()
   at Apache.Phoenix.WireMessage.MergeFrom(CodedInputStream input)
   at Google.Protobuf.MessageExtensions.MergeFrom(IMessage message, Stream input)
   at Google.Protobuf.MessageParser`1.ParseFrom(Stream input)
   at PhoenixSharp.PhoenixClient.<ExecuteRequestAsync>d__4a.MoveNext()
 --- End of inner exception stack trace ---
    at System.Threading.Tasks.Task.ThrowIfExceptional(Boolean includeTaskCanceledExceptions)
   at System.Threading.Tasks.Task.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at System.Threading.Tasks.Task.Wait()
   at Garuda.Data.PhoenixConnection.InternalExecuteRequest(PrepareResponse prepared, String sql, PhoenixParameterCollection parameterValues) in C:\Code\DotNet\GarudaUtil\Garuda.Data\PhoenixConnection.cs:line 267
   at Garuda.Data.PhoenixCommand.Execute() in C:\Code\DotNet\GarudaUtil\Garuda.Data\PhoenixCommand.cs:line 235
   at Garuda.Data.PhoenixCommand.ExecuteNonQuery() in C:\Code\DotNet\GarudaUtil\Garuda.Data\PhoenixCommand.cs:line 136
   at Garuda.Data.Test.PhoenixUnitTest.CommandPrepareWith1String1IntParamsTest() in C:\Code\DotNet\GarudaUtil\Garuda.Data.Test\PhoenixUnitTest.cs:line 285
Result Message:	
Test method Garuda.Data.Test.PhoenixUnitTest.CommandPrepareWith1String1IntParamsTest threw exception: 
System.AggregateException: One or more errors occurred. ---> Google.Protobuf.InvalidProtocolBufferException: While parsing a protocol message, the input ended unexpectedly in the middle of a field.  This could mean either than the input has been truncated or that an embedded message misreported its own length.
```

**Comment by brycehodson** [comment]
What does the command text you're giving to the PrepareRequestAsync() function look like, and what code are you using to create the parameter list?

Sending a non-string parameter should look very similar to sending a string parameter; you just need to set the TypedValue differently. For instance:
```
TypedValue tv1 = new TypedValue
{
    DoubleValue = 39.11,
    Type = Rep.DOUBLE
};
```

**Comment by dwdii** [comment]
Below is the command text I'm using. It is a test table with many data types. `:1` is a string value, and `:2` is a TINYINT.  Come to think of it, the tinyint is a bad choice for this test, though I've had the same problem with an int and bigint as I recall. I'll run another revised test to confirm the results.

Anyway, here is the command text:

```
UPSERT INTO GARUDATEST (ID, AircraftIcaoNumber, MyInt, MyUint, MyUlong, MyTingInt, MyTime, MyDate, MyTimestamp, MyUnsignedTime, MyFloat) VALUES (NEXT VALUE FOR garuda.testsequence, :1, 12, 14, 87, :2, CURRENT_TIME(), CURRENT_DATE(), '2016-07-25 22:28:00',  CURRENT_TIME(), 1.2 / .4)
```

The code to create the parameter list comes from a combination of my parameter collection wrapper and the parameter wrapper classes. The following code snippet is from the  [PhoenixParameter](https://github.com/dwdii/GarudaUtil/blob/master/Garuda.Data/PhoenixParameter.cs) class:
```
        public TypedValue AsPhoenixTypedValue()
        {
            var tv = new TypedValue();

            if(this.Value == null)
            {
                tv.Null = true;
            }
            else
            {
                Type pt = this.Value.GetType();
                if (pt == typeof(int) || pt == typeof(long))
                {
                    tv.NumberValue = Convert.ToInt64(this.Value);
                    tv.Type = Rep.BIG_INTEGER;
                }
                else if(pt == typeof(float) || pt == typeof(double))
                {
                    tv.DoubleValue = Convert.ToDouble(this.Value);
                    tv.Type = Rep.DOUBLE;
                }
                else if (pt == typeof(string))
                {
                    tv.StringValue = Convert.ToString(this.Value);
                    tv.Type = Rep.STRING;
                }
                else if (this.Value == DBNull.Value)
                {
                    tv.Null = true;
                    tv.Type = Rep.NULL;
                }
            }

            return tv;
        }
```

**Comment by dwdii** [comment]
My 1 string/1 int test (INT instead of TINYINT) resulted in the same exception:

```
Google.Protobuf.InvalidProtocolBufferException: While parsing a protocol message, the input ended unexpectedly in the middle of a field.  This could mean either than the input has been truncated or that an embedded message misreported its own length.
```

Interestingly, when I tried a 1 string/1 float variation, this succeeded.

I then added a test for a 1 string / 1 long (ulong in db) pair of parameters, and this failed with the InvalidProtocolBufferException also.

Finally - I considered my choice of Rep type for the TypedValue. I had been using `Rep.BIG_INTEGER`, but decided to try`Rep.LONG` and this was successful! As a result, my revised `AsPhoenixTypedValue` function now looks like the following: 

```{csharp}
       public TypedValue AsPhoenixTypedValue()
        {
            var tv = new TypedValue();

            if(this.Value == null)
            {
                tv.Null = true;
                tv.Type = Rep.NULL;
            }
            else
            {
                Type pt = this.Value.GetType();
                if (pt == typeof(int) || pt == typeof(long))
                {
                    tv.NumberValue = Convert.ToInt64(this.Value);
                    tv.Type = Rep.LONG;
                }
                else if(pt == typeof(float) || pt == typeof(double))
                {
                    tv.DoubleValue = Convert.ToDouble(this.Value);
                    tv.Type = Rep.DOUBLE;
                }
                else if (pt == typeof(string))
                {
                    tv.StringValue = Convert.ToString(this.Value);
                    tv.Type = Rep.STRING;
                }
                else if (this.Value == DBNull.Value)
                {
                    tv.Null = true;
                    tv.Type = Rep.NULL;
                }
            }

            return tv;
        }
```

So, in conclusion, `Rep.BIG_INTEGER` is not the same as `Rep.LONG` (although they are typically both 64 bit integers), and the ProtoBuf runtime throws the InvalidProtocolBufferException when the BIG_INTEGER is used where a LONG should be used.  Definitely not an issue with the Microsoft.Phoenix.Client, but arguably could be an issue with ProtoBuf.

**Comment by dwdii** [comment]
@duoxu and @brycehodson - Thanks for your replies. 

Looking at the PQS server log (/var/log/hbase/phoenix-hbase-server.log), it looks like the PQS Avatica version I have doesn't know about BIG_INTEGER yet. I'm running HDP 2.4.2 with HBase 1.1.2 and Phoenix 4.4 and I guess BIG_INTEGER is a newer Rep data type that Microsoft.Phoenix.Client knows about but my server doesn't. 

```
java.lang.IllegalArgumentException: No enum constant org.apache.calcite.avatica.ColumnMetaData.Rep.BIG_INTEGER
        at java.lang.Enum.valueOf(Enum.java:238)
        at org.apache.calcite.avatica.ColumnMetaData$Rep.valueOf(ColumnMetaData.java:422)
        at org.apache.calcite.avatica.ColumnMetaData$Rep.fromProto(ColumnMetaData.java:536)
        at org.apache.calcite.avatica.remote.TypedValue.fromProto(TypedValue.java:483)
        at org.apache.calcite.avatica.remote.Service$ExecuteRequest.deserialize(Service.java:1185)
        at org.apache.calcite.avatica.remote.Service$ExecuteRequest.deserialize(Service.java:1145)
        at org.apache.calcite.avatica.remote.RequestTranslator.transform(RequestTranslator.java:40)
        at org.apache.calcite.avatica.remote.ProtobufTranslationImpl.parseRequest(ProtobufTranslationImpl.java:252)
        at org.apache.calcite.avatica.remote.ProtobufHandler.decode(ProtobufHandler.java:42)
        at org.apache.calcite.avatica.remote.ProtobufHandler.decode(ProtobufHandler.java:28)
        at org.apache.calcite.avatica.remote.AbstractHandler.apply(AbstractHandler.java:95)
        at org.apache.calcite.avatica.remote.ProtobufHandler.apply(ProtobufHandler.java:38)
        at org.apache.calcite.avatica.server.AvaticaProtobufHandler.handle(AvaticaProtobufHandler.java:68)
        at org.eclipse.jetty.server.handler.HandlerList.handle(HandlerList.java:52)
        at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:97)
        at org.eclipse.jetty.server.Server.handle(Server.java:499)
        at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:311)
        at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:257)
        at org.eclipse.jetty.io.AbstractConnection$2.run(AbstractConnection.java:544)
        at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:635)
        at org.eclipse.jetty.util.thread.QueuedThreadPool$3.run(QueuedThreadPool.java:555)
        at java.lang.Thread.run(Thread.java:745)
```

Thanks again - definitely appreciate the help understanding the issue!

Best,

Daniel
